asm/exceptasm.s

```diff
@@ -69,646 +80,905 @@ glabel __osIntTable
+/**
+ * The exception preamble is copied to the exception vectors at
+ * UT_VEC, XUT_VEC, ECC_VEC, E_VEC, to direct execution to __osException
+ */
+LEAF(__osExceptionPreamble)
+    lui     $k0, %hi(__osException)
+    addiu   $k0, %lo(__osException)
+    jr      $k0
+     nop
+END(__osExceptionPreamble)
+
+LEAF(__osException)
+    # Load scratch space for thread saving
+    lui     $k0, %hi(__osThreadSave)
+    addiu   $k0, %lo(__osThreadSave)
+    # Save $at
+    sd      $at, THREAD_AT($k0)
+    # Save sr
+    mfc0    $k1, C0_SR
+    sw      $k1, THREAD_SR($k0)
+    # Disable interrupts
+    li      $at, ~(SR_IE | SR_EXL)
+    and     $k1, $k1, $at
+    mtc0    $k1, C0_SR
+    # Save some temp registers for use in the following
+    sd      $t0, THREAD_T0($k0)
+    sd      $t1, THREAD_T1($k0)
+    sd      $t2, THREAD_T2($k0)
+    # Mark FPU as unused
+    sw      $zero, THREAD_FP($k0)
+    # Left over from misplaced ifdef, immediately overwritten on next instruction
+    mfc0    $t0, C0_CAUSE
+savecontext:
+    # Save the previously running thread's context to be restored when it resumes
+    move    $t0, $k0
+    lui     $k0, %hi(__osRunningThread)
+    lw      $k0, %lo(__osRunningThread)($k0)
+    ld      $t1, THREAD_AT($t0)
+    sd      $t1, THREAD_AT($k0)
+    ld      $t1, THREAD_SR($t0)
+    sd      $t1, THREAD_SR($k0)
+    ld      $t1, THREAD_T0($t0)
+    sd      $t1, THREAD_T0($k0)
+    ld      $t1, THREAD_T1($t0)
+    sd      $t1, THREAD_T1($k0)
+    ld      $t1, THREAD_T2($t0)
+    sd      $t1, THREAD_T2($k0)
+    lw      $k1, THREAD_SR($k0)
+    mflo    $t0
+    sd      $t0, THREAD_LO($k0)
+    mfhi    $t0
+    andi    $t1, $k1, SR_IMASK
+    sd      $v0, THREAD_V0($k0)
+    sd      $v1, THREAD_V1($k0)
+    sd      $a0, THREAD_A0($k0)
+    sd      $a1, THREAD_A1($k0)
+    sd      $a2, THREAD_A2($k0)
+    sd      $a3, THREAD_A3($k0)
+    sd      $t3, THREAD_T3($k0)
+    sd      $t4, THREAD_T4($k0)
+    sd      $t5, THREAD_T5($k0)
+    sd      $t6, THREAD_T6($k0)
+    sd      $t7, THREAD_T7($k0)
+    sd      $s0, THREAD_S0($k0)
+    sd      $s1, THREAD_S1($k0)
+    sd      $s2, THREAD_S2($k0)
+    sd      $s3, THREAD_S3($k0)
+    sd      $s4, THREAD_S4($k0)
+    sd      $s5, THREAD_S5($k0)
+    sd      $s6, THREAD_S6($k0)
+    sd      $s7, THREAD_S7($k0)
+    sd      $t8, THREAD_T8($k0)
+    sd      $t9, THREAD_T9($k0)
+    sd      $gp, THREAD_GP($k0)
+    sd      $sp, THREAD_SP($k0)
+    sd      $fp, THREAD_S8($k0)
+    sd      $ra, THREAD_RA($k0)
+    beqz    $t1, savercp
+     sd     $t0, THREAD_HI($k0)
+    # If any CPU interrupts are enabled in the previous thread's SR, bitwise-OR in the
+    # disabled CPU interrupts from the global interrupt mask. The COP0 SR that is
+    # saved here is really the thread's SR masked by the global interrupt mask, so this
+    # is an attempt to recover the thread's SR from the current SR. This is however
+    # broken, see comments for osSetIntMask.
```

Comment:
> The COP0 SR that is saved here is really the thread's SR masked by the global interrupt mask
the "that is saved here" feels wrong, since what is actually saved has the `|`d bits

also I don't think the distinction between the "COP0 SR" (Sr the thread wants & global mask, I assume) and the "thread's SR" (Sr the thread wants, I assume) would make sense for another reader

maybe both these remarks can be addressed by providing less details and leaving it up to `osSetIntMask` to explain:

```suggestion
    # disabled CPU interrupts from the global interrupt mask.
    # This is an attempt at reverting the effect of bitwise-AND-ing the SR
    # with the global interrupt mask.
    # This is however broken, see comments for osSetIntMask.
```